src/util/address.rs

```diff
@@ -205,8 +205,125 @@ impl Payload {
+
+    /// Creates a pay to (compressed) public key hash address payload from a public key.
+    pub fn p2pkh(pk: &ecdsa::PublicKey) -> Payload {
+        let mut hash_engine = PubkeyHash::engine();
+        pk.write_into(&mut hash_engine)
+            .expect("engines don't error");
+        Payload::PubkeyHash(PubkeyHash::from_engine(hash_engine))
+    }
+
+    /// Creates a pay to script hash P2SH address payload from a script
+    #[inline]
+    pub fn p2sh(script: &script::Script) -> Payload {
+        Payload::ScriptHash(ScriptHash::hash(&script[..]))
+    }
+
+    /// Create a witness pay to public key address payload from a public key
+    pub fn p2wpkh(pk: &ecdsa::PublicKey) -> Result<Payload, Error> {
+        if !pk.compressed {
+            return Err(Error::UncompressedPubkey);
+        }
+
+        let mut hash_engine = WPubkeyHash::engine();
+        pk.write_into(&mut hash_engine)
+            .expect("engines don't error");
+        Ok(Payload::WitnessProgram {
+            version: bech32::u5::try_from_u8(0).expect("0<32"),
+            program: WPubkeyHash::from_engine(hash_engine)[..].to_vec(),
+        })
+    }
+
+    /// Create a pay to script address payload that embeds a witness pay to public key
+    pub fn p2shwpkh(pk: &ecdsa::PublicKey) -> Result<Payload, Error> {
+        if !pk.compressed {
+            return Err(Error::UncompressedPubkey);
+        }
+
+        let mut hash_engine = WPubkeyHash::engine();
+        pk.write_into(&mut hash_engine)
+            .expect("engines don't error");
+
+        let builder = script::Builder::new()
+            .push_int(0)
+            .push_slice(&WPubkeyHash::from_engine(hash_engine)[..]);
+
+        Ok(Payload::ScriptHash(ScriptHash::hash(
+            builder.into_script().as_bytes(),
+        )))
+    }
+
+    /// Create a witness pay to script hash address payload
+    pub fn p2wsh(script: &script::Script) -> Payload {
+        Payload::WitnessProgram {
+            version: bech32::u5::try_from_u8(0).expect("0<32"),
+            program: WScriptHash::hash(&script[..])[..].to_vec(),
+        }
+    }
+
+    /// Create a pay to script address payload that embeds a witness pay to script hash address
+    pub fn p2shwsh(script: &script::Script) -> Payload {
+        let ws = script::Builder::new()
+            .push_int(0)
+            .push_slice(&WScriptHash::hash(&script[..])[..])
+            .into_script();
+
+        Payload::ScriptHash(ScriptHash::hash(&ws[..]))
+    }
+}
+
+/// A utility struct to encode an address payload with the given parameters.
+pub struct AddressEncoding<'a> {
+    /// The address payload to encode.
+    pub payload: &'a Payload,
+    /// base58 version byte for p2pkh payloads (e.g. 0x00 for "1..." addresses).
+    pub p2pkh_prefix: u8,
+    /// base58 version byte for p2sh payloads (e.g. 0x05 for "3..." addresses).
+    pub p2sh_prefix: u8,
+    /// hrp used in bech32 addresss (e.g. "bc" for "bc1..." addresses).
+    pub bech32_hrp: &'a str,
```

Comment:
Might be best to have different lifetimes for `bech32_hrp` and `payload`. In most cases, the prefix would be `static` and other `'a` for payload.

I am not against creating an owned string copy here as it will be 2-3 characters max.

Reply:
I added two lifetimes 'a and 'b - imho no need to enforce `'static` here when the compiler can infer it or the user can choose the lifetime at the call site.

Reply:
Agreed

Reply:
I'm not convinced two lifetimes are useful. It's very rare that they are needed and this doesn't seem to be one of them.

Reply:
:shrug: I don't mind either way. Maybe @sanket1729 or @apoelstra have an example where it might be useful? Otherwise I can revert back to one lifetime.

Reply:
`&'a str` is covariant over `'a` so there's nothing preventing you assigning `'static` there. Multiple lifetimes are extremely rare. I recommend checking out this video on lifetimes: https://www.youtube.com/watch?v=rAl-9HwD858

Reply:
Sometimes I use to multiple lifetimes to aid reviews, they really help me understand what's going on with the code. I seldom use `&'wallet` `&'block` , `&'tx` in my code to help distinguish where the data comes from. 

It's really not any rule, just that I have gotten used to it annotating lifetimes when there are multiple references that make sense to annotate with meaningful names.

Reply:
Just stylistic preference and best left to author's coding style :)

Reply:
Interesting, never seen anything like it. Not convinced it's worth the boilerplate but I'm not capable of thinking well right now anyway. :)